File: core/BinaryTreeMaximumPathSum/BinaryTreeMaximumPathSum.java
package com.dbc;

public class BinaryTreeMaximumPathSum {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    private int res = -2000;
    private int dfs(TreeNode root) {
        if (root == null) return 0;
        int leftNum = 0, rightNum = 0, maxVal = 0;
        if (root.left != null) leftNum = dfs(root.left);
        if (root.right != null) rightNum = dfs(root.right);
        maxVal = Math.max(leftNum + root.val, rightNum + root.val);
        maxVal = Math.max(maxVal, root.val);
        this.res = Math.max(this.res, maxVal);
        this.res = Math.max(this.res, root.val + leftNum + rightNum);
        return maxVal;
    }

    public int maxPathSum(TreeNode root) {
        dfs(root);
        return this.res;
    }
}

File: core/BinaryTreeMaximumPathSum/BinaryTreeMaximumPathSum.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def maxPathSum(root: Optional[TreeNode]) -> int:
    res = -2000
    def dfs(node):
        if not node:
            return 0

        left_val, right_val = 0, 0
        if node.left:
            left_val = dfs(node.left)
        if node.right:
            right_val = dfs(node.right)

        max_val = max(left_val + node.val, right_val + node.val)
        max_val = max(max_val, node.val)
        nonlocal res
        res = max(res, max_val)
        res = max(res, left_val + right_val + node.val)
        return max_val

    dfs(root)
    return res





